<p>Shared naming conventions allow teams to collaborate efficiently. This rule checks that all function names match a provided regular expression.</p>
<h2>Noncompliant Code Example</h2>
<p>With default provided regular expression:</p>
<pre>
def DoSomething( ) : Unit = { // Noncompliant
   // ...
}
</pre>
<h2>Compliant Solution</h2>
<pre>
def doSomething( ) : Unit = {
   // ...
}
</pre>

