<p>The <code>match</code> statement should be used only to clearly define some new branches in the control flow. As soon as a <code>case</code> clause
contains too many statements this highly decreases the readability of the overall control flow statement. In such case, the content of the
<code>case</code> clause should be extracted into a dedicated method.</p>
<h2>Noncompliant Code Example</h2>
<p>With the default threshold of 5:</p>
<pre>
myVariable match {
  case 0 =&gt; // Noncompliant: 6 lines till next case
    methodCall1()
    methodCall2()
    methodCall3()
    methodCall4()
    methodCall5()
  case 1 =&gt;
  // ...
}
</pre>
<h2>Compliant Solution</h2>
<pre>
myVariable match {
  case 0 =&gt; doSomething()
  case 1 =&gt;
  // ...
}
// ...
def doSomething(): Unit = {
  methodCall1()
  methodCall2()
  methodCall3()
  methodCall4()
  methodCall5()
}
</pre>

