<p>For better readability, do not put more than one statement on a single line.</p>
<h2>Noncompliant Code Example</h2>
<pre>
println("Hello"); println("world!")
</pre>
<h2>Compliant Solution</h2>
<pre>
println("Hello")
println("world!")
</pre>

