<p>Nested <code>match</code> structures are difficult to understand because you can easily confuse the cases of an inner <code>match</code> as
belonging to an outer statement. Therefore nested <code>match</code> statements should be avoided.</p>
<p>Specifically, you should structure your code to avoid the need for nested <code>match</code> statements, but if you cannot, then consider moving
the inner <code>match</code> to another function.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def foo(n: Int, m: Int): Unit = {
  n match {
    case 0 =&gt; m match {
        case 0 =&gt;
        // ...
      }
    case 1 =&gt;
    // ...
  }
}
</pre>
<h2>Compliant Solution</h2>
<pre>
def foo(n: Int, m: Int): Unit = {
  n match {
    case 0 =&gt; bar(m)
    case 1 =&gt;
    // ...
  }
}

def bar(m: Int): Unit = {
  m match {
    case 0 =&gt;
    // ...
  }
}
</pre>

