<p>Merging collapsible <code>if</code> statements increases the codeâ€™s readability.</p>
<h2>Noncompliant Code Example</h2>
<pre>
if !filename.nil?
  if File.file?(filename) || File.directory?(filename)
    # ...
  end
end

if a then
  unless b then
    # ...
  end
end
</pre>
<h2>Compliant Solution</h2>
<pre>
def isFileOrDirectory(filename)
  File.file?(filename) || File.directory?(filename)
end
# ...

if !filename.nil? &amp;&amp; isFileOrDirectory(filename)
  # ...
end

if a &amp;&amp; !b then
  # ...
end
</pre>

