<p>Most of the time a block of code is empty when a piece of code is really missing. So such empty block must be either filled or removed.</p>
<h2>Noncompliant Code Example</h2>
<pre>
def compute(a, b)
  sum = a + b
  if  sum &gt; 0 # Noncompliant; empty on purpose or missing piece of code?
  end
  puts "Result: #{sum}"
end
</pre>
<h2>Compliant Solution</h2>
<pre>
def compute(a, b)
  sum = a + b
  if  sum &gt; 0
    puts "Positive result"
  end
  puts "Result: #{sum}"
end
</pre>
<h2>Exceptions</h2>
<p>When a block contains a comment, this block is not considered to be empty.</p>

