<p>The use of parentheses, even those not required to enforce a desired order of operations, can clarify the intent behind a piece of code. But
redundant pairs of parentheses could be misleading, and should be removed. </p>
<h2>Noncompliant Code Example</h2>
<pre>
x = (y / 2 + 1)  # Compliant even if the parenthesis are ignored by the compiler

if a &amp;&amp; ((x+y &gt; 0))  # Noncompliant
  # ...
end

return ((x + 1)) # Noncompliant
</pre>
<h2>Compliant Solution</h2>
<pre>
x = (y / 2 + 1)

if a &amp;&amp; (x+y &gt; 0)
  # ...
end

return (x + 1)
</pre>

