<h2>Why is this an issue?</h2>
<p>Nested <code>case</code> structures are difficult to understand because you can easily confuse the cases of an inner <code>case</code> as belonging
to an outer statement. Therefore nested <code>case</code> statements should be avoided.</p>
<p>Specifically, you should structure your code to avoid the need for nested <code>case</code> statements, but if you cannot, then consider moving the
inner <code>case</code> to another function.</p>
<h3>Noncompliant code example</h3>
<pre>
def foo(n, m)
  case n
  when 0
      case m  # Noncompliant; nested case
      when 0 then puts "0"
        # ...
      end
  when 1 then puts "1"
  else puts "2"
  end
end
</pre>
<h3>Compliant solution</h3>
<pre>
def foo(n, m)
  case n
  when 0
    bar(m)
  when 1 then puts "1"
  else puts "2"
  end
end

def bar(m)
  case m
  when 0 then puts "0"
    # ...
  end
end
</pre>

