<p>Having all branches in a <code>case</code> or <code>if</code> chain with the same implementation is an error. Either a copy-paste error was made
and something different should be executed, or there shouldn't be a <code>case</code>/<code>if</code> chain at all.</p>
<h2>Noncompliant Code Example</h2>
<pre>
if b == 0  # Noncompliant
  doOneMoreThing()
else
  doOneMoreThing()
end

b = a &gt; 12 ? 4 : 4;  # Noncompliant

case i  # Noncompliant
  when 1
    doSomething()
  when 2
    doSomething()
  when 3
    doSomething()
  else
    doSomething()
end
</pre>
<h2>Exceptions</h2>
<p>This rule does not apply to <code>if</code> chains or <code>case</code>-es without <code>else</code>-s.</p>
<pre>
if b == 0 # no issue, this could have been done on purpose to make the code more readable
  doSomething()
elsif b == 1
  doSomething()
end
</pre>

